core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java

```diff
@@ -145,40 +236,139 @@ public GeoBoundingBoxQueryBuilder ignoreMalformed(boolean ignoreMalformed) {
      * Sets the type of executing of the geo bounding box. Can be either `memory` or `indexed`. Defaults
      * to `memory`.
      */
-    public GeoBoundingBoxQueryBuilder type(String type) {
-        this.type = type;
+    public GeoBoundingBoxQueryBuilder type(Type type) {
+        this.type = (type != null) ? type : Type.MEMORY;
         return this;
     }
 
+    /** Returns the execution type of the geo bounding box.*/
+    public Type type() {
+        return type;
+    }
+
+    /** Returns the name of the field to base the bounding box computation on. */
+    public String fieldName() {
+        return this.fieldName;
+    }
+
+    /** Set to false to skip normlizing bounding box corners. */
+    public GeoBoundingBoxQueryBuilder normalize(boolean normalize) {
+        this.normalize = normalize;
+        return this;
+    }
+
+    /** Returns true if bounding box corners should be normalized. */
+    public boolean normalize() {
+        return this.normalize;
+    }
+
     @Override
-    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-        // check values
-        if(Double.isNaN(box[TOP])) {
-            throw new IllegalArgumentException("geo_bounding_box requires top latitude to be set");
-        } else if(Double.isNaN(box[BOTTOM])) {
-            throw new IllegalArgumentException("geo_bounding_box requires bottom latitude to be set");
-        } else if(Double.isNaN(box[RIGHT])) {
-            throw new IllegalArgumentException("geo_bounding_box requires right longitude to be set");
-        } else if(Double.isNaN(box[LEFT])) {
-            throw new IllegalArgumentException("geo_bounding_box requires left longitude to be set");
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (normalize == false) {
+            if (GeoUtils.isValidNumber(topLeft.getLat()) == false) {
+                validationException = addValidationError("Top part of top-left corner contains invalid number: " + topLeft.getLat(), validationException);
+            }
+            if (GeoUtils.isValidNumber(topLeft.getLon()) == false) {
+                validationException = addValidationError("Left part of top-left corner contains invalid number: " + topLeft.getLon(), validationException);
+            }
+            if (GeoUtils.isValidNumber(bottomRight.getLat()) == false) {
+                validationException = addValidationError("Bottom part of bottom-right corner contains invalid number: " + bottomRight.getLat(), validationException);
+            }
+            if (GeoUtils.isValidNumber(bottomRight.getLon()) == false) {
+                validationException = addValidationError("Top part of bottom-right corner contains invalid number: " + bottomRight.getLon(), validationException);
```

Comment:
in general, I wonder if we should be more explicit about whether the problem is in lon or lat, rather that saying e.g. "Top part of ..."